### ISSUE TITLE
Are there any plans for supporting an explicit attention mask?

### ISSUE THREAD

**Issue by Avelina9X** [Dao-AILab/flash-attention#840]
I've noticed that the Triton implementation supports explicit attention bias, which can be used to support arbitrary mask shapes with large negative values, however is there any planned support for explicit (boolean) masks in the CUDA implementation?

I've noticed some requests for features like off-diagonal attention, but an explicit attention mask would be able to facilitate this and any other arbitrary masking scheme - such as XL-Net, attention sinking, landmark attention - without needing to hardcode the attention scheme and enable it with an argument or seperate python interface.

**Comment by normster** [comment]
It seems like the [PyTorch attention implementation](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) supports custom attention masks and also uses Flash-Attention 2: https://twitter.com/StasBekman/status/1736083447658225665. Though I'm not sure that passing in an attention mask doesn't cause the op to dispatch to a non-FA2 kernel.

**Comment by tridao** [comment]
If there's attn mask pytorch does not dispatch to FA2 kernel, rather the kernel from xformers.

**Comment by abdulfatir** [comment]
Thanks for the info @tridao! Is support for arbitrary attention masks on your roadmap? This would be incredibly useful for some encoder-decoder and prefixLM models. Mandatory thank you for your amazing work!

**Comment by xiabingquan** [comment]
> If there's attn mask pytorch does not dispatch to FA2 kernel, rather the kernel from xformers.

Thanks for this valuable tip. No wonder torch.nn.functional.scaled_dot_product_attention does not bring any speed up in my case

**Comment by lin-ht** [comment]
I'm looking for bias mask support too, in FA2 and better FA3. Is there a roadmap for this? Thank you~

**Comment by nikonikolov** [comment]
Also very eager to see custom attn masks working. Seems like it was on the line 2 years ago, but it didn't get in. Is there a way to get such a feature in? Currently, there doesn't seem to be any alternative which combines flash attention speedups (especially on hopper) together with custom attention masks and this can massively speed up training some big models
